nurse_scheduling.py

```diff
@@ -0,0 +1,162 @@
+# K., Nakamura, Y. & Humble, T.S. Application of Quantum Annealing to Nurse
+# Scheduling Problem. Sci Rep 9, 12837 (2019).
+# https://doi.org/10.1038/s41598-019-49172-3, © The Author(s) 2019, use of
+# which is licensed under a Creative Commons Attribution 4.0 International
+# License (To view a copy of this license, visit
+# http://creativecommons.org/licenses/by/4.0/).
+
+from dwave.system import LeapHybridSampler
+from dimod import BinaryQuadraticModel
+from collections import defaultdict
+from copy import deepcopy
+
+# count nurses n = 1 ... n_nurses
+# count scheduling days as d = 1 ... n_days
+# binary variable q_nd is the assignment of nurse n to day d
+# a is a positive correlation coefficient for implementing the hard nurse
+# constraint - value provided by Ikeda, Nakamura, Humble
+a = 3.5
+n_nurses = 3
+n_days = 11
+size = n_days * n_nurses
+
+# Hard shift constraint: at least one nurse working every day
+# Lagrange parameter, for hard nurse constraint, on workforce and effort
+# Workforce function W(d) - set to a constant 'workforce' for now
+# Effort function E(n) - set to a constant 'effort' for now
+lagrange_parameter = 1.3
+workforce = 1
+effort = 1
+
+# Soft nurse constraint: all nurses should have approximately even work
+#                        schedules
+# Lagrange parameter, for shift constraints, on work days
+# preference function 'preference' - set to a constant for now
+# Minimum duty days 'min_duty_days' - the number of work days that each
+# nurse wants
+# to be scheduled. At present, each will do the minimum on average.
+# The parameter gamma's value suggested by Ikeda, Nakamura, Humble
+gamma = 0.3
+preference = 1
+min_duty_days = int(n_days/n_nurses)
+
+
+# Find index into 1D list for (nurse_index, day_index)
+def get_index(nurse_index, day_index):
+    return nurse_index * n_days + day_index
+
+
+# Inverse of get_index - given an index in a 1D list, return the nurse_index
+# and day_index
+def get_nurse_and_day(index):
+    nurse_index, day_index = divmod(index, n_days)
+    return nurse_index, day_index
+
+
+# Hard nurse constraint: no nurse works two consecutive days
+# It does not have Lagrange parameter - instead, J matrix
+# symmetric, real-valued interaction matrix J, whereas all terms are
+# a or zero.
+# composite indices i(n, d) and j(n, d) as functions of n and d
+# In other words, the indices are not n and d. They are kind of
+# (n+d)**2, and we need to take that into account. This seems to be
+# a clean way to do it.
+# J_i(n,d)j(n,d+1) = a and 0 otherwise.
+J = defaultdict(int)
+for nurse in range(n_nurses):
+    for day in range(n_days - 1):
+        nurse_day_1 = get_index(nurse, day)
+        nurse_day_2 = get_index(nurse, day+1)
+        J[nurse_day_1, nurse_day_2] = a
+
+# Q matrix assign the cost term, the J matrix
+Q = deepcopy(J)
+
+# Hard shift constraint. The sum is over each day.
+# lagrange_parameter * ((sum(effort * q) - workforce) ** 2)
```

Comment:
The equation in your comment here does not seem to match was is expressed in the code. Perhaps there is a way to clarify the relationship between the equation in your comment that of the code?

Things that I am confused by:
* your expression has a variable, `q`, but the code does not have such a variable
* currently `effort` is a scalar value. Is `effort * q` meant to be a matrix multiplication? If not, I don't see the point of `sum(effort * q)`
* The term `(sum(effort * q) - workforce) ** 2` is a square, yet I don't quite see it in the code. Even if you expanded out the expression, I would have expected to see some squared values. (eg. `workforce ** 2`). I understand that you took out `workforce**2` from the diagonal values, but what about the rest of the QUBO?

Reply:
@m3ller, I've updated the comments. The q is not a scalar, and I've indicated that. 'effort' is a scalar right now, in the paper. The algebra of squaring the terms is in the next several lines. See if this is better, and if not, we can discuss how to represent it in the code.

Reply:
@m3ller my algebra was wrong! It didn't affect the results, because the constants were equal to 1, but it was wrong. I've pushed what I hope is a fix.